### PR TITLE
Eric/dao/order

### PR DESCRIPTION
**Pull request by Eric-Spooner**
Eric/dao/menu is included

### CODE REVIEW

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Menu.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    public void setEntries(List<MenuEntry> entries) {
+        this.entries = entries;
+    }
+
+    /**
+     * Initialise the list, if its null and add the given entry
+     * @param entry (should not be null)
+     */
+    public void addEntry(MenuEntry entry){
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Menu.java]: remove (unused and buggy - if(entries==null){add})

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/MenuValidator.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        }
+        if (object.getIdentity() == null) {
+            throw new ValidationException("Identity must not be null");
+        }
+    }
+
+    private void checkForRequiredDataAttributesForCreateAndUpdate(Menu object) throws ValidationException {
+        if(object.getName() == null){
+            throw new ValidationException("Menu name must not be null");
+        }else if(object.getName() == ""){
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/MenuValidator.java]: remove the else

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/MenuValidator.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            throw new ValidationException("Menu name must not be empty");
+        }
+        if(object.getEntries() == null){
+            throw new ValidationException("Menu entries not be null");
+        }else if(object.getEntries().size() == 0){
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/MenuValidator.java]: remove the else

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/MenuValidator.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            throw new ValidationException("There should be at least one menu entry");
+        }else{
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/MenuValidator.java]: remove the else

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/MenuValidator.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            for(MenuEntry entry:object.getEntries()){
+                if(entry.getIdentity() == null) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/MenuValidator.java]: use Validator < MenuEntry >

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OrderValidator.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            throw new ValidationException("Tax must not be null");
+        }else if(order.getTax().compareTo(BigDecimal.valueOf(0)) == -1 || order.getTax().compareTo(BigDecimal.valueOf(1)) == 1) {
+            throw new ValidationException("Tax must be between 0 (included) and 1 (included)");
+        }
+        if(order.getMenuEntry() == null) {
+            throw new ValidationException("Menu Entry must not be null");
+        }else if(order.getMenuEntry().getIdentity() == null) {
+            throw new ValidationException("Menu Entry must have an identity");
+        }
+        if(order.getTable() == null) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OrderValidator.java]: use Validator < Table >

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OrderValidator.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            throw new ValidationException("Brutto must not be null");
+        }else if(order.getBrutto().equals(BigDecimal.valueOf(0))) {
+            throw new ValidationException("Brutto must not be 0");
+        }
+        if(order.getTax() == null) {
+            throw new ValidationException("Tax must not be null");
+        }else if(order.getTax().compareTo(BigDecimal.valueOf(0)) == -1 || order.getTax().compareTo(BigDecimal.valueOf(1)) == 1) {
+            throw new ValidationException("Tax must be between 0 (included) and 1 (included)");
+        }
+        if(order.getMenuEntry() == null) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OrderValidator.java]: use Validator < MenuEntry >

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/UserValidator.java [hunk 8]
```diff
@@ -58,7 +58,6 @@ private void checkForRequiredDataAttributesForCreateAndUpdate(User user) throws
         if (user.getName().isEmpty()) {
             throw new ValidationException("name must not be empty");
         }
-
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/UserValidator.java]: revert

File: src/main/resources/Spring-DAO.xml [hunk 9]
```diff
@@ -4,14 +4,16 @@
        xsi:schemaLocation="http://www.springframework.org/schema/beans
                            http://www.springframework.org/schema/beans/spring-beans.xsd">
 
-    <import resource="Spring-DataSource.xml" />
     <import resource="Spring-Domain.xml" />
+    <import resource="Spring-DataSource.xml" />
```
Comment (emmanuel099) [inline src/main/resources/Spring-DAO.xml]: revert

File: src/main/resources/Spring-Domain.xml [hunk 10]
```diff
@@ -13,6 +13,7 @@
     <import resource="domain/Spring-UserValidator.xml" />
     <import resource="domain/Spring-InvoiceValidator.xml"/>
     <import resource="domain/Spring-TaxRateValidator.xml" />
+    <import resource="domain/Spring-MenuValidator.xml" />
+    <import resource="domain/Spring-OrderValidator.xml" />
     <import resource="domain/Spring-MenuEntryValidator.xml" />
-
```
Comment (emmanuel099) [inline src/main/resources/Spring-Domain.xml]: revert

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestMenuValidator.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        // GIVEN
+        Menu menu = null;
+
+        // WHEN
+        menuValidator.validateForCreate(menu);
+    }
+
+
+    @Test(expected = ValidationException.class)
+         public void testValidateForCreate_MenuWithoutNameShouldThrow() throws ValidationException {
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestMenuValidator.java]: this test case is MenuWithEmptyName

add a additional test case which tests MenuWithoutName (means name == null)

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestOrderValidator.java [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+        order.setTime(LocalDateTime.now());
+        order.setAdditionalInformation("Info");
+        order.setInvoice(Invoice.withIdentity(1));
+
+        // WHEN
+        orderValidator.validateForCreate(order);
+    }
+
+    @Test(expected = ValidationException.class)
+         public void testValidateForCreate_OrderWithoutBruttoShouldThrow() throws ValidationException {
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestOrderValidator.java]: also test with negative brutto (should also throw)

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestOrderValidator.java [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+        order.setTime(LocalDateTime.now());
+        order.setAdditionalInformation("Info");
+        order.setInvoice(Invoice.withIdentity(1));
+
+        // WHEN
+        orderValidator.validateForCreate(order);
+    }
+
+    @Test(expected = ValidationException.class)
+    public void testValidateForCreate_OrderWithoutTaxShouldThrow() throws ValidationException {
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestOrderValidator.java]: also test out ouf bounds tax rate (one test case for tax rate < 0.0 and another one for tax rate > 1.0) (have a look at tax rate validator)

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestTaxRateValidator.java [hunk 14]
```diff
@@ -1,182 +0,0 @@
-package com.at.ac.tuwien.sepm.ss15.edulium.domain;
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestTaxRateValidator.java]: revert

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 15]
```diff
@@ -0,0 +1,577 @@
...
+                        list.add(newEntry.getIdentity());
+                    }
+            }
+            String queryByMenuEntry =
+                    "SELECT m.ID, m.name FROM (Menu m  JOIN MenuAssoc ma ON m.id  = ma.menu_id) WHERE " +
+                        "m.deleted = false AND ma.disabled = false AND " +
+                        "m.ID = ISNULL(?, ID) AND " +
+                        "m.name = ISNULL(?, name) AND " +
+                        "ma.menuEntry_ID in (" +
+                        getQuestionMarks(list.size()) +
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: list.stream().map(m -> "?").collect(Collectors.joining(", "))

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 16]
```diff
@@ -0,0 +1,577 @@
...
+            //at first find the MenuEntries, which have been given in the menu
+            List<Long> list = new LinkedList<>();
+            for(MenuEntry entry: menu.getEntries()) {
+                    List<MenuEntry> innerList = menuEntryDAO.find(entry);
+                    for(MenuEntry newEntry: innerList){
+                        list.add(newEntry.getIdentity());
+                    }
+            }
+            String queryByMenuEntry =
+                    "SELECT m.ID, m.name FROM (Menu m  JOIN MenuAssoc ma ON m.id  = ma.menu_id) WHERE " +
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: use a semi join instead of a normal join

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 17]
```diff
@@ -0,0 +1,577 @@
...
+            if (key.next()) {
+                menu.setIdentity(key.getLong(1));
+            }
+            key.close();
+        } catch (SQLException e) {
+            LOGGER.error("inserting menu into database failed", e);
+            throw new DAOException("inserting menu into database failed", e);
+        }
+        Long changeNr =  generateHistoryMenu(menu);
+        for(MenuEntry entry:menu.getEntries()) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: use batched sql updates instead of calling it one by one

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 18]
```diff
@@ -0,0 +1,577 @@
...
+                LOGGER.error("updating menu in database failed, dataset not found");
+                throw new DAOException("updating menu in database failed, dataset not found");
+            }
+        } catch (SQLException e) {
+            LOGGER.error("updating menu in database failed", e);
+            throw new DAOException("updating menu in database failed", e);
+        }
+        Long changeNr = generateHistoryMenu(menu);
+        /**
+         * Update Entries:
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: what if you disable all associations first, then insert/re-enable the associations from the menu via merge into?

have a look at DBReservationDAO.updateReservationAssociations

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 19]
```diff
@@ -0,0 +1,470 @@
...
+                LOGGER.error("deleting menu failed, dataset not found");
+                throw new DAOException("deleting menu failed, dataset not found");
+            }
+        } catch (SQLException e) {
+            LOGGER.error("deleting menu failed", e);
+            throw new DAOException("deleting menu failed", e);
+        }
+        Long changeNr = generateHistoryMenu(menu);
+
+        for (MenuEntry entry : menu.getEntries()) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: not needed

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 20]
```diff
@@ -0,0 +1,470 @@
...
+            deleteMenuAssoc(menu, entry);
+            generateHistoryMenuAssoc(menu, entry, changeNr);
+        }
+
+        menu.setEntries(new LinkedList<>());
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: remove

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 21]
```diff
@@ -0,0 +1,470 @@
...
+        if (menu == null) {
+            return new ArrayList<>();
+        }
+
+        final List<Menu> objects = new ArrayList<>();
+
+        //Check if the user also wants to search for MenuEntries
+        if (menu.getEntries() != null && menu.getEntries().size() > 0) {
+            //at first find the MenuEntries, which have been given in the menu
+            List<Long> list = new LinkedList<>();
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: List<MenuEntry>

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 22]
```diff
@@ -0,0 +1,470 @@
...
+            for (MenuEntry entry : menu.getEntries()) {
+                List<MenuEntry> innerList = menuEntryDAO.find(entry);
+                for (MenuEntry newEntry : innerList) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: use list.addAll(menuEntryDAO.find(entry)); instead

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 23]
```diff
@@ -0,0 +1,470 @@
...
+                    list.add(newEntry.getIdentity());
+                }
+            }
+            String queryByMenuEntry =
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: final

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 24]
```diff
@@ -0,0 +1,470 @@
...
+                    "SELECT m.ID, m.name FROM (Menu m JOIN MenuAssoc ma ON m.id  = ma.menu_id) WHERE " +
+                            "m.deleted = false AND ma.disabled = false AND " +
+                            "m.ID = ISNULL(?, ID) AND " +
+                            "m.name = ISNULL(?, name) AND " +
+                            "ma.menuEntry_ID in (" +
+                            list.stream().map(m -> "?").collect(Collectors.joining(",")) +
+                            ") " +
+                            "Group By m.ID";
+
+            objects.clear();
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: remove

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 25]
```diff
@@ -0,0 +1,470 @@
...
+            //at first find the MenuEntries, which have been given in the menu
+            List<Long> list = new LinkedList<>();
+            for (MenuEntry entry : menu.getEntries()) {
+                List<MenuEntry> innerList = menuEntryDAO.find(entry);
+                for (MenuEntry newEntry : innerList) {
+                    list.add(newEntry.getIdentity());
+                }
+            }
+            String queryByMenuEntry =
+                    "SELECT m.ID, m.name FROM (Menu m JOIN MenuAssoc ma ON m.id  = ma.menu_id) WHERE " +
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: replace this join by a semi join 

select * from Menu m where m.disabled = false AND ... AND exists (select 1 from MenuAssoc ma where m.id = ma.menu_id AND ma.disabled = false)

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 26]
```diff
@@ -0,0 +1,470 @@
...
+                ResultSet result = stmt.executeQuery();
+                while (result.next()) {
+                    objects.add(parseResult(result));
+                }
+            } catch (SQLException e) {
+                LOGGER.error("searching for menu entries failed", e);
+                throw new DAOException("searching for menu entries failed", e);
+            }
+        } else { //Only search for the name and/or the id
+            String query = "SELECT * FROM Menu WHERE " +
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: final

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 27]
```diff
@@ -0,0 +1,470 @@
...
+            while (resultEntries.next()) {
+                MenuEntry entry = MenuEntry.withIdentity(resultEntries.getLong("menuEntry_ID"));
+                menuEntries.add(menuEntryDAO.find(entry).get(0));
+            }
+        } catch (SQLException e) {
+            LOGGER.error("searching for menu entries failed", e);
+            throw new DAOException("searching for menu entries failed", e);
+        }
+
+        if (menuEntries.size() > 0) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: empty list is also valid

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 28]
```diff
@@ -0,0 +1,426 @@
...
+
+            if (stmt.executeUpdate() == 0) {
+                LOGGER.error("deleting menu failed, dataset not found");
+                throw new DAOException("deleting menu failed, dataset not found");
+            }
+        } catch (SQLException e) {
+            LOGGER.error("deleting menu failed", e);
+            throw new DAOException("deleting menu failed", e);
+        }
+        menu.setEntries(new LinkedList<>());
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: remove (delete must not change the menu)

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 29]
```diff
@@ -0,0 +1,426 @@
...
+        String query = "SELECT * FROM MenuAssoc WHERE " +
+                "menu_ID = ? AND " +
+                "disabled = false";
+        try (PreparedStatement stmt = dataSource.getConnection().prepareStatement(query)) {
+            stmt.setObject(1, result.getLong("ID"));
+
+            ResultSet resultEntries = stmt.executeQuery();
+            while (resultEntries.next()) {
+                MenuEntry entry = MenuEntry.withIdentity(resultEntries.getLong("menuEntry_ID"));
+                menuEntries.add(menuEntryDAO.find(entry).get(0));
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: will throw a out of bounds exception when find returns a empty list

check if the result of menuEntryDAO.find(entry) is of size 1, if not throw a dao exception

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 30]
```diff
@@ -0,0 +1,426 @@
...
+            LOGGER.error("generating Menu history failed", e);
+            throw new DAOException("generating Menu history failed", e);
+        }
+        //get the change Nr, for Menu Assoc
+        final String queryGetChangeNr = "SELECT MAX(changeNr) FROM MenuHistory WHERE ID = ?";
+        Long resultNr = -1L;
+        try (PreparedStatement stmt = dataSource.getConnection().prepareStatement(queryGetChangeNr)) {
+            stmt.setLong(1, menu.getIdentity());          // dataset id
+            ResultSet result = stmt.executeQuery();
+            while (result.next()) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: if (result.next()) {
   resultNr = result.getLong(1)
} else {
   throw
}

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOrderDAO.java [hunk 31]
```diff
@@ -0,0 +1,316 @@
...
+        LOGGER.debug("entering update with parameters " + order);
+
+        validator.validateForUpdate(order);
+
+        final String query = "UPDATE RestaurantOrder SET invoice_ID = ?, table_section = ?, table_number = ?," +
+                " menuEntry_ID = ?, orderTime = ?, brutto = ?, tax = ?, info = ? WHERE ID = ?";
+
+        try (PreparedStatement stmt = dataSource.getConnection().prepareStatement(query)) {
+            if(order.getInvoice() == null){
+                stmt.setNull(1, Types.VARCHAR);
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOrderDAO.java]: Types.LONG or BIGINT ?

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOrderDAO.java [hunk 32]
```diff
@@ -0,0 +1,316 @@
...
+                stmt.setNull(6, Types.VARCHAR);
+                stmt.setNull(7, Types.VARCHAR);
+            }
+            else {
+                stmt.setObject(6, order.getInvoice().getIdentity());
+                stmt.setObject(7, order.getInvoice().getIdentity());
+            }
+            stmt.setObject(8, order.getTable() == null ? null : order.getTable().getSection() == null ?
+                    null : order.getTable().getSection().getIdentity());
+            stmt.setObject(9, order.getTable() == null ? null : order.getTable().getNumber());
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOrderDAO.java]: if (order.getTable() == null) {
    stmt.setNull(8)
   stmt.setNull(9)
} else {
  Table table = order.getTable();
  stmt.setNull(table.getSection() == null ? null : table.getSection().getIdentity())
   stmt.setLong(table.getNumber())
}

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java [hunk 33]
```diff
@@ -0,0 +1,438 @@
...
+                    "SELECT m.ID, m.name FROM Menu m WHERE " +
+                            "m.deleted = false AND " +
+                            "m.ID = ISNULL(?, ID) AND " +
+                            "m.name = ISNULL(?, name) AND " +
+                            "EXISTS (SELECT 1 FROM MenuAssoc ma WHERE " +
+                            "ma.menu_ID = m.ID AND "+
+                            "ma.menuEntry_ID in (" +
+                            menu.getEntries().stream().map(m -> "?").collect(Collectors.joining(",")) +
+                            ") AND ma.disabled = false)" +
+                            "Group By m.ID";
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBMenuDAO.java]: group by isn't needed anymore
